src/drivers/dshot/DShot.h

```diff
@@ -0,0 +1,234 @@
+#endif
+
+/** Dshot PWM frequency, Hz */
+static constexpr unsigned int DSHOT150  =  150000u;
+static constexpr unsigned int DSHOT300  =  300000u;
+static constexpr unsigned int DSHOT600  =  600000u;
+static constexpr unsigned int DSHOT1200 = 1200000u;
+
+static constexpr int DSHOT_DISARM_CMD   = dshot_command_t::DShot_cmd_motor_stop;
+static constexpr int DSHOT_MIN_THROTTLE = 1;
+static constexpr int DSHOT_MAX_THROTTLE = 1999;
+
+class DShot : public cdev::CDev, public ModuleBase<DShot>, public OutputModuleInterface
+{
+public:
+	DShot();
+	virtual ~DShot();
+
+	enum Mode {
+		MODE_NONE = 0,
+		MODE_1PWM,
+		MODE_2PWM,
+		MODE_2PWM2CAP,
+		MODE_3PWM,
+		MODE_3PWM1CAP,
+		MODE_4PWM,
+		MODE_4PWM1CAP,
+		MODE_4PWM2CAP,
+		MODE_5PWM,
+		MODE_5PWM1CAP,
+		MODE_6PWM,
+		MODE_8PWM,
+		MODE_14PWM,
+		MODE_4CAP,
+		MODE_5CAP,
+		MODE_6CAP,
+	};
+
+	/** Mode given via CLI */
+	enum PortMode {
+		PORT_MODE_UNSET = 0,
+		PORT_FULL_GPIO,
+		PORT_FULL_PWM,
+		PORT_PWM8,
+		PORT_PWM6,
+		PORT_PWM5,
+		PORT_PWM4,
+		PORT_PWM3,
+		PORT_PWM2,
+		PORT_PWM1,
+		PORT_PWM3CAP1,
+		PORT_PWM4CAP1,
+		PORT_PWM4CAP2,
+		PORT_PWM5CAP1,
+		PORT_PWM2CAP2,
+		PORT_CAPTURE,
+	};
+
+	static void capture_trampoline(void *context, const uint32_t channel_index, const hrt_abstime edge_time,
+				       const uint32_t edge_state, const uint32_t overflow);
+
+	/** @see ModuleBase */
+	static int custom_command(int argc, char *argv[]);
+
+	Mode get_mode() { return _mode; }
+
+	virtual int init();
+
+	virtual int ioctl(file *filp, int cmd, unsigned long arg);
+
+	/** change the mode of the running module */
+	static int module_new_mode(const PortMode new_mode);
+
+	void mixerChanged() override;
+
+	/** @see ModuleBase::print_status() */
+	int print_status() override;
+
+	/** @see ModuleBase */
+	static int print_usage(const char *reason = nullptr);
+
+	void retrieve_and_print_esc_info_thread_safe(const int motor_index);
+
+	/**
+	 * Send a dshot command to one or all motors
+	 * This is expected to be called from another thread.
+	 * @param num_repetitions number of times to repeat, set at least to 1
+	 * @param motor_index index or -1 for all
+	 * @return 0 on success, <0 error otherwise
+	 */
+	int send_command_thread_safe(const dshot_command_t command, const int num_repetitions, const int motor_index);
+
+	int set_mode(const Mode new_mode);
+
+	/** @see ModuleBase */
+	static int task_spawn(int argc, char *argv[]);
+
+	bool telemetry_enabled() const { return _telemetry != nullptr; }
+
+	bool updateOutputs(bool stop_motors, uint16_t outputs[MAX_ACTUATORS],
+			   unsigned num_outputs, unsigned num_control_groups_updated) override;
+
+private:
+
+	/** Disallow copy construction and move assignment. */
+	DShot(const DShot &) = delete;
+	DShot operator=(const DShot &) = delete;
+
+	enum class DShotConfig {
+		Disabled  = 0,
+		DShot150  = 150,
+		DShot300  = 300,
+		DShot600  = 600,
+		DShot1200 = 1200,
+	};
+
+	struct Command {
+		const char *name{nullptr};
```

Comment:
Why did you add this?

Reply:
so that this could be condensed: https://github.com/PX4/PX4-Autopilot/blob/25c113a5273ab315c511773b23ef81e2a93f0026/src/drivers/dshot/dshot.cpp#L1408

Reply:
Please don't. `Command` is used internally, and might get filled in in future from other sources as well. `VerbCommand` is specifically for CLI, so the separation is on purpose.

Reply:
Removed!  Thanks @bkueng !